### PR TITLE
make it possible to create and reference nodes within a single transaction

### PR DESCRIPTION
**Pull request by lsmith77**
see https://github.com/phpcr/phpcr-api-tests/pull/165 and https://github.com/jackalope/jackalope/pull/297

related to https://github.com/jackalope/jackalope-doctrine-dbal/pull/292

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
     },
     "require-dev": {
         "psr/log": "~1.0",
-        "phpcr/phpcr-api-tests": "2.1.5",
+        "phpcr/phpcr-api-tests": "dev-create_and_reference_nodes_in_transaction",
```
Comment (lsmith77) [inline composer.json]: change to 2.1.6 once phpcr/phpcr-api-tests#165 is merged and tagged

**Comment by dbu** [comment]
when trying this scenario with the JavaDavexClient and jackrabbit 2.9, i get an UnsupportedRepositoryOperationException because the value factory tries to get the uuid of a non-persisted node. 

but if we can set uuids from the client, we could discover this case and create a uuid for the referenced node rather than let the backend do that.

**Comment by lsmith77** [comment]
yeah that was my plan. in phpcr odm we set uuids early. so it should be possible.

**Comment by dbu** [comment]
in phpcr-utils ValueConverter, we use getIdentifier, which we found out that for non-referenceable nodes can be the path. we might want to change Node::getIdentifier() to check itself for being of type mix:referenceable when there is no jcr:uuid property and create the uuid in that case. instead of simply returning the path even as new but referenceable node.  tried to find where exactly this happens in the dbal client that we translate the path to a uuid but did not find it.

**Comment by dbu** [comment]
ah, you already found it in dbal: https://github.com/jackalope/jackalope-doctrine-dbal/pull/292/files#r36081252 - i think if we make the Node::getIdentifier create a uuid if needed, the changes in https://github.com/jackalope/jackalope-doctrine-dbal/pull/292 could be partially reverted (i don't like too much to mix path and uuid in that way, things become rather implicit and detecting mistakes would become harder).

**Comment by dbu** [comment]
we will need to figure out where to put the uuid generator. maybe into the jackalope session? has the client access to the session to use it there? a custom generator would be written if the uuid generator needs to interact with the client to generate the uuid.

oh, and once we have new nodes generate a uuid, we could probably get rid of the uuid generator in phpcr-odm and just use $node->getIdentifier() at https://github.com/doctrine/phpcr-odm/blob/master/lib/Doctrine/ODM/PHPCR/UnitOfWork.php#L1781 and part of https://github.com/doctrine/phpcr-odm/blob/master/lib/Doctrine/ODM/PHPCR/UnitOfWork.php#L3748-L3765

**Comment by lsmith77** [comment]
that sounds like a very good idea. i think it would simplify things in the doctrine dbal PR

**Comment by lsmith77** [comment]
probably just requires updating some mock object setup to be able to support calls to ``isNodeType()``

File: composer.json [hunk 2]
```diff
@@ -21,14 +21,14 @@
         "ext-curl":"*",
         "phpcr/phpcr": "~2.1.2",
         "phpcr/phpcr-utils": "^1.2.8",
-        "jackalope/jackalope": "~1.2.3"
+        "jackalope/jackalope": "dev-ensure_uuid_is_set as 1.2.3"
```
Comment (dbu) [inline composer.json]: when merged and tagged, update this line

File: composer.json [hunk 3]
```diff
@@ -21,14 +21,14 @@
...
     },
     "provide": {
         "jackalope/jackalope-transport": "1.1.0"
     },
     "require-dev": {
         "psr/log": "~1.0",
-        "phpcr/phpcr-api-tests": "2.1.9",
+        "phpcr/phpcr-api-tests": "dev-create_and_reference_nodes_in_transaction as 2.1.9",
```
Comment (dbu) [inline composer.json]: when merged and tagged update this line

**Pull request closed without merging**